src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -65,52 +67,234 @@ ucs_status_t uct_rdmacm_cm_reject(struct rdma_cm_id *id)
     return UCS_OK;
 }
 
-ucs_status_t uct_rdmacm_cm_get_cq(uct_rdmacm_cm_t *cm, struct ibv_context *verbs,
-                                  struct ibv_cq **cq_p)
+static ucs_status_t
+uct_rdmacm_cm_device_context_init(uct_rdmacm_cm_device_context_t *ctx,
+                                  uct_rdmacm_cm_t *cm,
+                                  struct ibv_context *verbs)
+{
+    const char *dev_name = ibv_get_device_name(verbs->device);
+
+#if HAVE_DECL_MLX5DV_IS_SUPPORTED
+    char out[UCT_IB_MLX5DV_ST_SZ_BYTES(query_hca_cap_out)] = {};
+    char in[UCT_IB_MLX5DV_ST_SZ_BYTES(query_hca_cap_in)]   = {};
+    uint64_t general_obj_types_caps;
+    void *cap;
+    int ret;
+
+    if (cm->config.reserved_qpn == UCT_RDMACM_CM_RESERVED_QPN_NO) {
+        goto dummy_qp_ctx_init;
+    }
+
+    if (!mlx5dv_is_supported(verbs->device)) {
+        ucs_debug("%s: mlx5dv is not supported", dev_name);
+        goto dummy_qp_ctx_init;
+    }
+
+    cap = UCT_IB_MLX5DV_ADDR_OF(query_hca_cap_out, out, capability);
+    UCT_IB_MLX5DV_SET(query_hca_cap_in, in, opcode,
+                      UCT_IB_MLX5_CMD_OP_QUERY_HCA_CAP);
+    UCT_IB_MLX5DV_SET(query_hca_cap_in, in, op_mod,
+                      (UCT_IB_MLX5_CAP_GENERAL << 1) |
+                      UCT_IB_MLX5_HCA_CAP_OPMOD_GET_CUR);
+    ret = mlx5dv_devx_general_cmd(verbs, in, sizeof(in),
+                                  out, sizeof(out));
+    if (ret != 0) {
+        ucs_debug("mlx5dv_devx_general_cmd(%s, QUERY_HCA_CAP) failed: %m",
+                  dev_name);
+        goto dummy_qp_ctx_init;
+    }
+
+    general_obj_types_caps =
+            UCT_IB_MLX5DV_GET64(cmd_hca_cap, cap, general_obj_types);
+    if (!(general_obj_types_caps & UCS_BIT(UCT_IB_MLX5_OBJ_TYPE_RESERVED_QPN))) {
+        ucs_debug("%s general_obj_types_caps: reserved qpn is not support", dev_name);
+        goto dummy_qp_ctx_init;
+    }
+    
+    UCT_IB_MLX5DV_SET(query_hca_cap_in, in, op_mod,
+                      (UCT_IB_MLX5_CAP_2_GENERAL << 1) |
+                      UCT_IB_MLX5_HCA_CAP_OPMOD_GET_CUR);
+    ret = mlx5dv_devx_general_cmd(verbs, in, sizeof(in),
+                                  out, sizeof(out));
+    if (ret != 0) {
+        ucs_debug("mlx5dv_devx_general_cmd(%s, QUERY_HCA_CAP_2) failed: %m", dev_name);
+        goto dummy_qp_ctx_init;
+    }
+
+    ctx->log_reserved_qpn_granularity =
+            UCT_IB_MLX5DV_GET(cmd_hca_cap_2, cap, log_reserved_qpn_granularity);
+    ucs_debug("%s reserved qpn cap: log_reserved_qpn_granularity is 0x%x",
+              dev_name, ctx->log_reserved_qpn_granularity);
+
+    ctx->reserved_qpn_support = 1;
```

Comment:
maybe rename 'reserved_qpn_support' to 'use_reserved_qpn'?
since now it could be supported, but not enabled (by configuration)